akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Behaviors.scala

```diff
@@ -282,20 +282,6 @@ object Behaviors {
   @DoNotInherit
   trait Receive[T] extends ExtensibleBehavior[T]
 
-  /**
-   * Per message MDC (Mapped Diagnostic Context) logging.
-   *
-   * @param mdcForMessage Is invoked before each message is handled, allowing to setup MDC, MDC is cleared after
-   *                 each message processing by the inner behavior is done.
-   * @param behavior The actual behavior handling the messages, the MDC is used for the log entries logged through
-   *                 `ActorContext.log`
-   *
-   * See also [[akka.actor.typed.Logger.withMdc]]
-   */
-  def withMdc[T](
-    mdcForMessage: akka.japi.function.Function[T, java.util.Map[String, Any]], behavior: Behavior[T]): Behavior[T] =
-    WithMdcBehavior[T](Map.empty, message ⇒ mdcForMessage.apply(message).asScala.toMap, behavior)
-
```

Comment:
I thought the 3 overloads were overkill, and they are bad for type inference (and compilation error messages).

Reply:
I see that you mean with the overloads but I think they all pull their weight hm

Reply:
To clarify, the problem is for example:

```
      def behavior3: Behavior[String] =
        Behaviors.withMdc[String](msg ⇒ Map("a" -> msg)) {
          Behaviors.receive {
            case (ctx, msg) ⇒
              ctx.log.info(msg.toUpperCase())
              Behaviors.same
          }
        }

[error] /Users/patrik/dev/akka/akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorLoggingSpec.scala:323:27: missing parameter type
[error]         Behaviors.withMdc(msg ⇒ Map("a" -> msg)) {
```

One would have to add `(msg: String) =>`

It only goes away completely if removing all overloads and only keeping
```
def withMdc[T](staticMdc: Map[String, Any], mdcForMessage: T ⇒ Map[String, Any])(behavior: Behavior[T]): Behavior[T]
```

So my removal of one overloads is not worth much, reverting it and creating a separate ticket.

Reply:
I played around with giving the static one a separate name `withStaticMdc` earlier today but didn't see any improvement, but maybe that is the path to go down three separate names?

Reply:
yeah, having separate names would solve it (then not overloading), however how big api surface do we really need for mdc? Then it should start with same `withMdc...` but I can't come up with any good such naming

Reply:
`withMdcStatic`, `withMdcPerMessage`, `withMdc` doesn't sound so bad or crazy api-surface-wise to me.

Reply:
Created issue https://github.com/akka/akka/issues/24798, with another suggestion